# Reading PDF files in Python
* Reading text data from PDF files
* Extract page-wise text content in a PDF file
* Create corpus using text data from multiple PDF files 

In [1]:
pwd

'/Users/sylvia/Desktop/IITR/M6_text_analytics/Times_Demo_Videos'

In [2]:
# !pip install -U -q PyDrive
# from pydrive.auth import GoogleAuth
# from pydrive.drive import GoogleDrive
# from google.colab import auth
# from oauth2client.client import GoogleCredentials

In [3]:
# auth.authenticate_user()
# gauth = GoogleAuth()
# gauth.credentials = GoogleCredentials.get_application_default()
# drive = GoogleDrive(gauth)

In [4]:
# downloaded = drive.CreateFile({'id':'1FjJzaLd-qxFTv61J1oZNLpTtL5oYH7xI'}) # replace the id with id of file you want to access
# downloaded.GetContentFile('CommunicationAffectAndLearning.pdf') 

In [5]:
# !pip install PyPDF2

PyPDF2 to read PDF content in Python.

In [6]:
import PyPDF2


# path = 'CommunicationAffectAndLearning.pdf'
# pdf = PyPDF2.PdfFileReader(open(path, 'rb')) 

# Invoking a pdf file reader; using open to read the file in binary using 'rb'    


Open a file and get its handler. File opened in 'rb' mode which is read binary mode.

In [7]:
path = '/Users/sylvia/Desktop/datasets/CommunicationAffectAndLearning.pdf'
pdf = PyPDF2.PdfFileReader(open(path, 'rb')) 

#Invoking a pdf file reader; using open to read the file in binary using 'rb'

Xref table not zero-indexed. ID numbers for objects will be corrected.


In [8]:
# using an attribute numPages to know number of pages in pdf
pdf.numPages

267

In [9]:
# Using a method to get the info of the doc
info = pdf.getDocumentInfo()
print(info.author)
print(info.title)

Owner
Microsoft Word - Communication Affect and Learning.doc


### Extract the text and read

- We read the pdf file by pages.
- This extractText() returns a byte object that's y u see b' written in front of any text you read.

In [10]:
# To read the first page

page = pdf.getPage(0)
# View as text
text = page.extractText().encode('utf-8') # extracting and encoding
text

# We see first page is empty

b''

In [11]:
page = pdf.getPage(1)   # 2nd pg
# View as text
text = page.extractText().encode('utf-8') # extracting and encoding
text

b'      Communication, Affect, & Learning in the Classroom \n   Jason S. Wrench Virginia Peck Richmond Joan Gorham \n '

In [12]:
page = pdf.getPage(2)   # 3rd pg
# View as text
text = page.extractText().encode('utf-8') # extracting and encoding
text

b' Copyright \xc2\xa9 2009 by Virginia Peck Richmond, Jason S. Wrench, and Joan Gorham All rights reserved. Printed in the United States of America.    This work is licensed under the Creative Commons Attribution-Noncommercial-No Derivative Works 3.0 United States License.   To view a copy of this license, visit http://creativecommons.org/licenses/by-nc-nd/3.0/us/ or send a letter to Creative Commons, 171 Second Street, Suite 300, San Francisco, California, 94105, USA.   All research measures utilized within this textbook are copyrighted by the measure\xe2\x80\x99s original creators and used by permission within this text. Please contact the measure\xe2\x80\x99s original creator for licensing information.               1st Edition printed by Burgess Publishing, Edina, MN, in 1992 (ISBN: 0-80874-699-5) 2\nnd Edition printed by Tapestry Press, Acton, MA, in 2001 (ISBN: 1-56888-548-2)'

page.extractText() returns a byte object.

In [13]:
# page.extractText().encode('utf-8') return a byte object
type(text)

bytes

### Convert the byte object to string type using function decode so u can do string operations on top of the text data.

decode function to convert bytes to string format.

In [14]:
page = pdf.getPage(2)
text = page.extractText().encode('utf-8').decode('utf-8')
print(type(text))
print('\n',text)

<class 'str'>

  Copyright © 2009 by Virginia Peck Richmond, Jason S. Wrench, and Joan Gorham All rights reserved. Printed in the United States of America.    This work is licensed under the Creative Commons Attribution-Noncommercial-No Derivative Works 3.0 United States License.   To view a copy of this license, visit http://creativecommons.org/licenses/by-nc-nd/3.0/us/ or send a letter to Creative Commons, 171 Second Street, Suite 300, San Francisco, California, 94105, USA.   All research measures utilized within this textbook are copyrighted by the measure’s original creators and used by permission within this text. Please contact the measure’s original creator for licensing information.               1st Edition printed by Burgess Publishing, Edina, MN, in 1992 (ISBN: 0-80874-699-5) 2
nd Edition printed by Tapestry Press, Acton, MA, in 2001 (ISBN: 1-56888-548-2)


## Read whole pdf at once - each page will be a document/string. 

In [15]:
read_pages = lambda pdf, pg_num : pdf.getPage(pg_num).extractText().encode('utf-8').decode('utf-8') 

# lambda fn taking two parameters(pdf is filehandler, pg_num is page no which we want to read) 
# lambda fn uses filehandler(pdf) to take page no and extract text out of it after converting 
# it into a string format

pages = [read_pages(pdf, x) for x in range(pdf.numPages)]    # List comprehension

In [16]:
len(pages)

267

So we have 267 strings. Each page is a string now.

In [17]:
pages[3]

' \n TABLE OF CONTENTS \n P r e f a c e            i   1.  Teaching As a Communication Process      0 0 1   The Instructional Communication Process  The Teacher  The Content  The Instructional Strategy  The Student  The Feedback/Evaluation   The Learning Environment/Instructional Context  Kibler’s Model of Instruction  The ADDIE Model of Instructional Design  2. Communicating With Instructional Objectives    017  Why Some Teachers Resent Objectives  The Value of Objectives  What Objectives Should Communicate  3. Instructional Communication Strategies     028  The Teacher As a Speaker  The Teacher As a Moderator  The Teacher As a Trainer  The Teacher As a Manager  The Teacher As a Coordinator & Innovator   4.  Communication, Affect, and Student Needs     042  Measuring Student Affect  Basic Academic Needs of Students  Traditional Interpersonal Need Models   Outcomes of Meeting Student Needs  5 .   L e a r n i n g  S t y l e s          0 5 9   What is Learning Style? '

In [18]:
pages[15]

' \nChapter One - 3 riffs, and putting them together made really fine music. His friends find this very impressive, and Spike can\'t wait to learn more.  What have these three piano teachers communicated to their students? Grady\'s teacher taught him to dislike studying piano, that playing the piano was hard work and no fun. Grady probably also learned that studying music is a pain and should be avoided in the future. He may avoid trying out for the band or chorus in school, and he may have already learned to associate music itself with unpleasantness so that he will not even choose to listen to much music as he grows older. Since it is very unlikely that these were the objectives of Grady\'s teacher, we must question that teacher\'s ability to utilize the instructional communication process effectively. Roxanne\'s teacher was somewhat more effective. Roxanne learned that music can be fun and enjoyable, but she also learned that she had very little musical ability. Her self-esteem as a

## Join multiples pages
* sometimes we have to use multiple pdf's so in that case each pdf content will be available in a single string to form a document and another string coming from another pdf to form 2nd document.
* So above although every page is a document (1 page as 1 string) because there is a single pdf however if there are multiple pdf then each pdf should be a 1 document (1 pdf as 1 string) then 658326 words would be a single string.
* Every document will always come in form of a single string.

In [19]:
one_pdf = ' '.join(pages) # Combining all the pages together using join fn by providing it a space; 
                          #Every page will have and additional space b/w them.   
len(one_pdf)

644684

One pdf will act like a single document in case u r reading multiple pdf's together (can be called a corpus)

# Read all PDFs and create corpus

In [20]:
# !pip install -U -q PyDrive
# from pydrive.auth import GoogleAuth
# from pydrive.drive import GoogleDrive
# from google.colab import auth
# from oauth2client.client import GoogleCredentials

In [21]:
# auth.authenticate_user()
# gauth = GoogleAuth()
# gauth.credentials = GoogleCredentials.get_application_default()
# drive = GoogleDrive(gauth)

In [22]:
#  # replace the id with id of file you want to access
# downloaded = drive.CreateFile({'id':'12E3xUHTnEgkX3z-PrX4bGzDHiJshydv4'})
# downloaded.GetContentFile('How to Measure Anything.pdf') 

In [23]:
# !pip install -U -q PyDrive
# from pydrive.auth import GoogleAuth
# from pydrive.drive import GoogleDrive
# from google.colab import auth
# from oauth2client.client import GoogleCredentials

In [24]:
# auth.authenticate_user()
# gauth = GoogleAuth()
# gauth.credentials = GoogleCredentials.get_application_default()
# drive = GoogleDrive(gauth)

In [25]:
# downloaded = drive.CreateFile({'id':'1xmcwhXmHHT5S9cMo_iYjltY1UUFx5JpM'})
# downloaded.GetContentFile('everythingfinance.pdf') 

***.pdf referring to all pdf's present in the folder.**

In [26]:
import glob # used for getting all the pdfs filenames from drive 
path_search = '/Users/sylvia/Desktop/datasets/books/*.pdf'
path_pdf_files = glob.glob(path_search)
path_pdf_files

['/Users/sylvia/Desktop/datasets/books/everythingfinance.pdf',
 '/Users/sylvia/Desktop/datasets/books/How to Measure Anything.pdf',
 '/Users/sylvia/Desktop/datasets/books/CommunicationAffectAndLearning.pdf']

# Create a corpus

- We have multiple pdf's so each pdf will be a document (a single string).
- Collection of documents will be a corpus.
- docs corpus below will have all individual pdf's content in the form of strings.

In [27]:
docs = [] # creating empty list named docs (it is like corpus), list will contain 3 strings/docs; Each pdf is 1 string
read_pages1 = lambda pdf, pg_num : pdf.getPage(pg_num).extractText().encode('utf-8').decode('utf-8') 



# lambda fn taking two parameters(pdf is filehandler, pg_num is page no which we want to read) 
# lambda fn uses filehandler(pdf) to take page no and extract text out of it after converting it into a string format
for x in path_pdf_files:
  pdf = PyPDF2.PdfFileReader(open(x, 'rb'))
  pages = [read_pages1(pdf, i) for i in range(pdf.numPages)]
  doc = '\n'.join(pages)
  docs.append(doc)


/Users/sylvia/opt/anaconda3/envs/TF1/lib/python3.7/site-packages/PyPDF2/_cmap.py:136: PdfReadWarning: Advanced encoding /Identity#2dH not implemented yet
  PdfReadWarning,
Xref table not zero-indexed. ID numbers for objects will be corrected.


Each pdf will be a single string which will be available in this docs list.

In [28]:
docs[0][0:500]      # 1st pdf 

'Everything You Ever Really Needed to Know About Personal FinanceOn Just One Page\nby Trent Hamm - The Simple Dollar - http://www.thesimpledollar.com/\n\nIntroductionThe cover of this document tells you the whole story.  Everything you really need to know abut personal finance can be summarized in just one page.  Spend less than you earn.  Earn more.  Live frugal.  Do something sensible with the difference.  Control your own destiny.  All of the other writing out there on personal finance is just deta'

In [29]:
docs[1][0:500]      # 2nd pdf

'\n\nHow to Measure\nAnything\nWorkbook\n\nDouglas W. Hubbar DHow to Measure\nAnything\nWorkbook\nFinding the Value of  “Intangibles”   in Business\nCover Design: Wiley\nCover Images, © iStockphoto.com (clockwise from the top); © graphxarts, © elly99, \n© derrrek, © procurator, © Olena_T, © miru5\nCopyright © 2014 by Douglas W. Hubbard. All rights reserved.\nPublished by John Wiley & Sons, Inc., Hoboken, New Jersey.\nFirst edition published by John Wiley & Sons, Inc. in 2007. \nSecond edition published by John W'